### DISCUSSION TITLE
EventPriority < 0 cause listener not fired

### DISCUSSION THREAD

**Discussion by ylazy** [airsdk/Adobe-Runtime-Support#1531] [General]
Please run this code:

```
var s:Sprite = new Sprite();
	s.addEventListener(Event.ADDED_TO_STAGE, addedToStage, false, int.MAX_VALUE, true);
	s.addEventListener(Event.ADDED_TO_STAGE, addedToStage2); // Not Fired!!!!!
	stage.addChild(s);
	
function addedToStage(e:Event):void
{
	trace("addedToStage");
	s.removeEventListener(Event.ADDED_TO_STAGE, addedToStage);
	s.addEventListener(Event.REMOVED_FROM_STAGE, removedFromStage, false, int.MIN_VALUE, true); // If you replace int.MIN_VALUE by 0, then addedToStage2 will be fired
}

function addedToStage2(e:Event):void
{
	trace("addedToStage2");
}

function removedFromStage(e:Event):void
{
	s.removeEventListener(Event.REMOVED_FROM_STAGE, removedFromStage);
}
```
You will see the listener addedToStage2 will not be fired.
If you replace:
`s.addEventListener(Event.REMOVED_FROM_STAGE, removedFromStage, false, int.MIN_VALUE, true);`
by
`s.addEventListener(Event.REMOVED_FROM_STAGE, removedFromStage, false, 0, true);`
then everything work fine.
But any value lower than 0, for example:
`s.addEventListener(Event.REMOVED_FROM_STAGE, removedFromStage, false, -1, true);`
will cause the listener addedToStage2 will not be fired.

I've attached a .fla file for quick testing.
[EventPriorityBug.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/7819468/EventPriorityBug.zip)

ASDoc:
```
/**
 * @param	priority	The priority level of the event listener. The priority is designated by
 *   a signed 32-bit integer. The higher the number, the higher the priority. All listeners
 *   with priority n are processed before listeners of priority n-1. If two
 *   or more listeners share the same priority, they are processed in the order in which they
 *   were added. The default priority is 0.
 */
public function addEventListener (type:String, listener:Function, useCapture:Boolean=false, priority:int=0, useWeakReference:Boolean=false) : void;
```

Please Help!
Thanks!

**Comment by ajwfrost** [comment]
Thanks for this -> looks like the addition of a new event listener with a different priority was causing the internal list iteration to go wrong. Fairly simple fix for this so we'll get this updated in the next release.

thanks

**Reply by toysonlocation** [reply]
If this is the cause, please keep it working :)

https://github.com/airsdk/Adobe-Runtime-Support/discussions/1321

**Reply by ajwfrost** [reply]
:-) no it's actually an issue that's existed from the very start, from what I can see... there's an iterator that goes through the different priority levels, and an iterator that goes through the listeners that are registered at each priority level. Adobe had already protected against changes to the second list i.e. within a single priority level, but if any handler changed the set of lists by adding something at a new priority level, it was possible for the code to skip over some other listeners.

What actually happens:
- event is triggered
- we have two sets of lists: priority 0 and priority int_max, the lists are stored in an ordered list
- we go through the int_max handlers which injects a new list at priority int_min
- our ordered list now has three sets of lists: priority int_min, 0 and int_max
- but we now have an invalid iterator variable and we look at the listeners on the "int_min" priority rather than the "0" priority
- ... and this doesn't include the expected handler, so it's not triggered
Fix is to detect the change and to re-start the iterator from the point at which we had previously left it...
